File: java/src/test/java/algorithm/tests/utils/ErrorsTest.java
package algorithm.tests.utils;

import static org.junit.jupiter.api.Assertions.assertEquals;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import algorithm.tests.fixtures.Helper;
import algorithm.utils.errors.parameters.NegativeParameterError;
import algorithm.utils.errors.parameters.ZeroParameterError;

@DisplayName("Errors Test Suite")
public class ErrorsTest {
    @Test
    @DisplayName("Negative Value Error Test")
    void negativeValuError_returnsCorrectMessage() {
        String name = "dimension";
        double value = -2.0;

        NegativeParameterError negative = new NegativeParameterError(name, value);

        String actual = negative.toString();
        String expected = "NegativeParameterError: DIMENSION is -2.0, but it must be a non-negative value.";

        assertEquals(expected, actual, Helper.errorMessage(actual, expected));
    }

    @Test
    @DisplayName("Zero Value Error Test")
    void zeroValueError_returnsCorrectMessage() {
        String name = "dimension";
        double value = 0.0;

        ZeroParameterError zero = new ZeroParameterError(name, value);

        String actual = zero.toString();
        String expected = "ZeroParameterError: DIMENSION is 0.0, but it must be a nonzero value.";

        assertEquals(expected, actual, Helper.errorMessage(actual, expected));
    }
}

File: rust/tests/utils/errors_test.rs
use algorithm::utils::errors::parameters::{NegativeParameterError, ZeroParameterError};

use crate::fixtures::helper;

#[test]
fn negative_value_error_returns_correct_message() {
    let name = "dimension".to_string();
    let value = -2.0;

    let negative: NegativeParameterError = NegativeParameterError { name, value };

    let actual = format!("{negative}");
    let expected = "NegativeParameterError: DIMENSION is -2, but it must be a positive value.";

    assert_eq!(actual, expected, "{}", helper::error_message(&actual, &expected.to_string()));
}

#[test]
fn test_zero_value_error_returns_correct_message() {
    let name = "dimension".to_string();
    let value = 0.0;

    let zero: ZeroParameterError = ZeroParameterError { name, value };

    let actual = format!("{zero}");
    let expected = "ZeroParameterError: DIMENSION is 0, but it must be a nonzero value.";

    assert_eq!(actual, expected, "{}", helper::error_message(&actual, &expected.to_string()));
}
